[win32][xbmc] Upgrade to VS2015

Paxxi opened the pull request:
This is currently marked as a WIP, it runs but it's not tested at all really.

The required libraries are rebuilt and available on the mirrors for testing.

I decided to remove the disabling of iterator debugging as it makes it easier to build libraries, less things to remember and less things to change in their build configs. This triggered some warnings of out of bounds access on empty arrays that are now checked for.

Might still be some ugliness in the fixes as the first goal was to get it running and testable, have begun doing some of the cleanup required.

There's also plenty of overlap here with the #8991 which will disappear once those are merged.

Paxxi committed:
Visual studio 2015 fixes

Paxxi commented:
@Razzeee @afedchin @Montellese @MartijnKaijser 
Feel free to try it out and see how it behaves.

afedchin commented on project/VS2010Express/XBMC.defaults.props:
```diff
@@ -2,12 +2,12 @@
     <IncludePath>$(SolutionDir)\..\BuildDependencies\include;$(SolutionDir)\..\BuildDependencies\include\python;$(IncludePath)</IncludePath>
     <ExecutablePath>$(SolutionDir)\..\..\tools\win32buildtools;$(ExecutablePath)</ExecutablePath>
-    <LocalDebuggerEnvironment>KODI_HOME=$(SolutionDir)..\..&#xD;&#xA;PATH=$(SolutionDir)..\Win32BuildSetup\dependencies;%PATH%</LocalDebuggerEnvironment>
+    <LocalDebuggerEnvironment>KODI_HOME=$(SolutionDir)..\..
```
looks like ; is missed after ..\..

Paxxi replied:
good catch

MartijnKaijser replied:
Does this need to be fixed before its forgotten?

Paxxi replied:
It's not broken, I tried fixing it and that broke it so I reverted the change but probably forgot to document it somewhere so this should not be touched.

afedchin replied:
seems vs2013 escapes line break with `&#xD;&#xA;` but vs2013 doesn't

afedchin commented:
also neet to include visual c++ redistributable 2015 in installer

Paxxi commented:
it's already included because sqlite is using it.

afedchin commented:
didn't know that.

Paxxi committed:
Fixed missing iterator include and broken environment setting in project files

Paxxi committed:
Fixed crashing in emu crt wrapper

Don't use memset on objects! horrible things happen
playing around with constexpr, probably more than necessary but might as
well learn to use it.
Apparently there shouldn't be a ; in the vc project environment settings

garbear committed:
[DllLoader] Fix compile on linux

Razzeee commented:
See https://github.com/Razzeee/xbmc/tree/vs2015 for an rebased version
And http://jenkins.kodi.tv/view/Windows/job/WIN-32/7984/ for a build test as we should now have the needed tools on zags windows slave (camelot might need the tools too?

Razzeee commented:
@MartijnKaijser confirmed, that camelot doesn't need to be updated.
So we should be good to merge this?

Voyager1 commented:
will this require everyone who builds on Windows to switch over to VS2015?

afedchin commented:
@Voyager1 yes, it requires

Razzeee commented:
jenkins build this please

Razzeee commented:
@fetzerch might this cause problems for the out of tree build?

fetzerch commented:
@Razzeee: If you mean the CMake build, yes that does require adaptation. From what I can see this is needed:
- `project/cmake/scripts/windows/archsetup.cmake`: remove _HAS_ITERATOR_DEBUGGING, add mysql to the dynloaded libs
- Adapt `project/cmake/modules/FindXXX.cmake` for the libs where we now have different libraries for debug and release.

It'll unfortunately take a bit until I can upgrade VisualStudio on my old windows box. Maybe some-when next week. Would be cool if we could get the Windows Jenkins build script updated to support CMake, than we'd immediately see what is broken.

Since you mentioned out-of-source builds: with CMake you can have a build directory that is different to the source directory and CMake then takes care to copy the relevant files (for example skins, system folder, ...) but as you don't add additional files, I think that should still be good.

Razzeee commented on project/BuildDependencies/scripts/0_package.list:
```diff
@@ -7,8 +7,8 @@
-crossguid-8f399e8bd4-win32.7z
-curl-7.42.1-win32.7z
+crossguid-8f399e-win32-vc140.7z
+curl-7.46.0-win32-vc140.7z
```
@mapfau Is this new enough?

mapfau replied:
I was not sure if this one can be used because it has to linked statically (compared to the older one wich contains the dll) - Refering to the DLLCurl implementation kodi currently expects dynamic linkage...

Razzeee replied:
No clue sorry, @paxxi likely knows

afedchin replied:
@mapfau is right DllLibCurl expects a dll in system folder. And we doesn't link libcurl at all.

mapfau replied:
I could try to build a package for it but this would take a lot of time because I have to setup all the things / directories and read and compare things. I guess that somebody who did this already before and has the environment for this could do this task in much shorter time and energy !!??

afedchin replied:
I'll try to do that if @Paxxi will not arise :)

Paxxi replied:
I can rebuild with latest curl, openssl is used by something else as well so that will be kept on current version. 

We have used statically linked curl before and it works fine in my testing

Paxxi committed:
Fix build and update curl to latest

Paxxi commented:
@mapfau built with latest libcurl, not runtime tested, my dev machine is in a semi-broken state currently. files should be on mirrors in ~30min

mapfau commented:
I simply replaced libcurl.dll in xbmc/system folder.
.lib and .exp files are not used at all (I renamed them to be sure)

everything i done in DLLCurl.h/cpp file with resolves the function pointer when its used.

really appreciate the work wich is done - to be honest: I dont want to switch to 2015
I could build the libCurl on 2013 if somebody shows em the way how to do....

Paxxi commented:
@mapfau there is a short Readme included in the curl zip, follow it and replace `c:\code\kdeps` with the path to project /buildependencies and it should build

mapfau commented:
just reread the thread and understood that curl is not the reason why we have to switch to 2015.
Sorry for confuing, thanx @paxxi for building it!

MartijnKaijser commented:
Can't load python27.dll and zlib.dll

mapfau commented:
i remember that I also had those naming problems - I renamed the zlib1.dll to zlib.dll (knowing that this is an ugly hack) I'll try to find out what the background is for that (if something has changed in naming conventions the near past)

garbear commented on project/BuildDependencies/scripts/0_package.list:
```diff
@@ -40,12 +38,14 @@ libxml2-2.7.8_1-win32.7z
+openssl-1.0.2e-win32-vc140.7z
+pcre-8.37-win32-vc140.7z
+pillow-3.1.0-win32-vc140.7z
+python-2.7.11-win32-vc140.7z
```
This causes Kodi to crash on start when calling `Py_Initialize();`. Replacing python27.dll with the old version fixes the crash

Razzeee replied:
Any info on the crash? Might be an addon if your system is not vanilla?

Montellese replied:
`Py_Initialize()` is a standard call of the Python API to set up the python interpreter and is what we have to call before running any python addon. If it crashes, something is wrong with the DLL.

garbear replied:
Kodi dies a violent death. all threads crash and return 0x01 the second this is called. probably a problem with the DLL

ace20022 replied:
This reminds me of  a similar crash I had months ago. The problem back then was that a python dll from the PATH was loaded. I don't think that this is the case here, but who knows.

ace20022 replied:
And again ;) ```'Kodi.exe' (Win32): Loaded 'C:\Program Files (x86)\<--->\python27.dll'.```

Paxxi replied:
I noticed I missed to add python and sqlite to delay load when building with cmake, will fix tonight

garbear commented:
@paxxi: fixed a compile bug on linux: https://github.com/Paxxi/xbmc/pull/1

DaveTBlake commented:
Since this will require everyone who builds on Windows to switch over to VS2015, can we take care over when this gets merged. Not looking forward to updating my dev setup, and would hate to rebase something only to find I'm stuck!

Paxxi commented:
@DaveTBlake install vs2015 community today and you're ready, they coexist perfectly fine, no need to to uninstall 2013 or touch anything else

garbear commented:
Right, I have both installed and it's painless to switch between the two

Razzeee commented:
I actually already use vs2015, but compiling with the 2013 compiler

Razzeee commented:
So can we merge this soon, as we're still pre alpha right now? There are a bunch of packages bumped, so it might be nice to do it early.

MartijnKaijser commented:
Not if it doesn't work or drop the problem parts to another PR

garbear commented:
ContextMenuManager.cpp needs an `#include <iterator>`

garbear commented on project/BuildDependencies/scripts/0_package.list:
```diff
@@ -40,12 +38,14 @@ libxml2-2.7.8_1-win32.7z
 texturepacker-1.1.0-win32.7z
-tinyxml-2.6.2_3-win32-vc120.7z
+tinyxmlstl-2.6.2-win32-vc140.7z
+zlib-1.2.8-win32-vc140-v2.7z
```
This is missing `system/zlibd.dll`, which was present in `zlib-1.2.8-win32-vc140.7z` (sans `v2.7`)

Paxxi replied:
It's not missing, I decided to skip the debug version as it's only really needed for statically linked libs

garbear committed:
Fix build for VS2015 update

garbear commented:
@paxxi: https://github.com/Paxxi/xbmc/pull/2

Paxxi commented:
Will try and have a look at the python issue tomorrow, last I ran this it worked fine but will investigate

Paxxi committed:
Fix debug build after switching to zlib dll instead of statically linked

Paxxi committed:
Use correct build tools for addons and get rid of old unused build file

Paxxi commented:
no python issues in debug mode, addons working fine.

I noticed issues with zlib that's hopefully fixed and appraently I had forgotten to switch the addons to use 2015 as well. Switched over now and triggered plenty of build errors in addons so this is some way away from being ready

Paxxi commented:
@Montellese or @fetzerch maybe you can help track down one remaining issue, when running buildsetup.bat I get this
```
[ 62%] Performing build step for 'modplug'
[  2%] Building CXX object CMakeFiles/modplug.dir/src/load_abc.cpp.obj
load_abc.cpp
C:\code\kodi\project\cmake\addons\build\modplug\src\modplug\src\load_abc.cpp(32): fatal error C1083: Cannot open include file: 'unistd.h': No such file or directory
```
Looking at the patch files included for modplug there should be no unistd.h include in that file so my guess is that the patches aren't applied for some reason but I don't really know where to start to track it down.
One clue is that it works with this branch before the last commit switching the addons to vs2015 but not sure how that could cause this.

fetzerch committed:
[cmake] VS2015 support

Paxxi commented:
jenkins build this please

Paxxi committed:
Updated appveyor to use vs2015 as well

Paxxi committed:
Fixed some missed paths

MartijnKaijser commented:
jenkins build this please

MartijnKaijser commented on xbmc/utils/CharsetConverter.cpp:
```diff
@@ -708,7 +713,7 @@ bool CCharsetConverter::wToUtf32(const std::wstring& wStringSrc, std::u32string&
 // The bVisualBiDiFlip forces a flip of characters for hebrew/arabic languages, only set to false if the flipping
 // of the string is already made or the string is not displayed in the GUI
-bool CCharsetConverter::utf8ToW(const std::string& utf8StringSrc, std::wstring& wStringDst, bool bVisualBiDiFlip /*= true*/,
+bool CCharsetConverter::utf8ToW(const std::string& utf8StringSrc, std::wstring& wStringDst, bool bVisualBiDiFlip /*= true*/, 
```
Unneeded change?

MartijnKaijser commented on xbmc/utils/CharsetConverter.cpp:
```diff
@@ -496,7 +501,7 @@ bool CCharsetConverter::CInnerConverter::logicalToVisualBiDi(const std::u32strin
     else
       lineEnd++; // include '\n'
-
+    
```
Unneeded white space changes (also above)?

Paxxi replied:
oh yeah those snuck in because I tried som VS addon to remove whitespace on save, in one of the earlier commits so too lazy to revert the changes. No harm done other than some noise in the diff

fetzerch commented on lib/win32/Effects11/Effects11_2013.vcxproj:
```diff
@@ -1,5 +1,5 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+<Project DefaultTargets="Build" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
There's btw also a Effects11_2015 solution and project. Shouldn't we use that instead?

Paxxi replied:
that could be a good idea, had no idea there was such a project in there, will have a look if there's any differences and possibly switch over

mapfau commented:
At what (build) step should sqlite3.dll be copied to wich place?
For sure I could copy the one from the python folder somwhere in my path environment, but I don't think this is what was planned.

mapfau commented:
I'm not sure if switching to VS2015 was such a good idea. I really hope, that nobody els has the same trouble like me. Its really so annoying:

```
C:\Dev\xbmc\project\cmake\addons\build>cmake -DADDONS_TO_BUILD="inputstream.mpd" -DADDON_SRC_PREFIX="C:\Dev\kodi_addons" -DCMAKE_BUILD_TYPE=Debug -G "Visual Studio 12"  -DCMAKE_USER_MAKE_RULES_OVERRIDE="C:\Dev\xbmc\project\cmake\scripts\windows\c-flag-overrides.cmake" -DCMAKE_USER_MAKE_RULES_OVERRIDE_CXX="C:\Dev\xbmc\project\cmake\scripts\windows\cxx-flag-overrides.cmake" -DCMAKE_INSTALL_PREFIX="C:\Dev\xbmc\addons" -DBUILD_DIR="C:\Dev\xbmc\project\cmake\addons\build" -DPACKAGE_ZIP=1 "C:\Dev\xbmc\project\cmake\addons"
-- The C compiler identification is unknown
-- The CXX compiler identification is unknown
CMake Error at CMakeLists.txt:1 (project):
  No CMAKE_C_COMPILER could be found.



CMake Error at CMakeLists.txt:1 (project):
  No CMAKE_CXX_COMPILER could be found.



-- Configuring incomplete, errors occurred!
See also "C:/Dev/xbmc/project/cmake/addons/build/CMakeFiles/CMakeOutput.log".
See also "C:/Dev/xbmc/project/cmake/addons/build/CMakeFiles/CMakeError.log".

C:\Dev\xbmc\project\cmake\addons\build>cmake -DADDONS_TO_BUILD="inputstream.mpd" -DADDON_SRC_PREFIX="C:\Dev\kodi_addons" -DCMAKE_BUILD_TYPE=Debug -G "Visual Studio 12"  -DCMAKE_USER_MAKE_RULES_OVERRIDE="C:\Dev\xbmc\project\cmake\scripts\windows\c-flag-overrides.cmake" -DCMAKE_USER_MAKE_RULES_OVERRIDE_CXX="C:\Dev\xbmc\project\cmake\scripts\windows\cxx-flag-overrides.cmake" -DCMAKE_INSTALL_PREFIX="C:\Dev\xbmc\addons" -DBUILD_DIR="C:\Dev\xbmc\project\cmake\addons\build" -DPACKAGE_ZIP=1 "C:\Dev\xbmc\project\cmake\addons"
-- The C compiler identification is unknown
-- The CXX compiler identification is unknown
CMake Error at CMakeLists.txt:1 (project):
  No CMAKE_C_COMPILER could be found.



CMake Error at CMakeLists.txt:1 (project):
  No CMAKE_CXX_COMPILER could be found.



-- Configuring incomplete, errors occurred!
See also "C:/Dev/xbmc/project/cmake/addons/build/CMakeFiles/CMakeOutput.log".
See also "C:/Dev/xbmc/project/cmake/addons/build/CMakeFiles/CMakeError.log".

C:\Dev\xbmc\project\cmake\addons\build>
```

Anybody an idea whats wrong here when building binary addons?

Thanks!

Paxxi commented:
Project/buildependencies/downloadbuilddeps will download sqlite and store the dll in system 
Not sure about the cmake stuff, have you cleaned the cmake cache/build folder before running with -G "visual Studio 14"?

mapfau commented:
thanx, @Paxxi , will try more

mapfau commented:
Following I found out over the day yesterday:
- cmake 3.1.1 is not able (at least on my system) to build addons.
After upgrading to the latest 3.5 cmake I was able to build addons.
- I'm still unable to build kodi with 8.1 windows kit, because my visual studio doesn't find sdk 10 folders wich are necessary (stdio.h for example, wich was in older VS versions inside VS but has moved to SDK10 now). I spend some more hours searching and many people have this problem with VS2015. Seems to be something with old entries in the registry.
- I copied finally sqlite.dll manually into the kodi build folder, but also other dlls are mising.
Question for this: what do I have to delete to force  downloadbuilddeps to rebuild everything from scratch? I did git clean -xf yesterday, but it doesn't work.

Thanks!

afedchin commented:
@mapfau Have you latest VS update installed? Yesterday I've installed vs2015 comunity (it's actualy vs2015 Update2) and was be able to compile and run Kodi.

mapfau commented:
I have already learned that its working fine with a fresh VS install - because of some reasons I have 4 VS versions installed (VS2010/13 express, VS2013/2015 Community). This all together seems to be not working very well.
I'll make an new 2015 insstallation today, yesterday I id a repair, but it took 4 hours with nothing fixing.

afedchin commented:
@mapfau I have vs2010/2012/2013/2015 installed

mapfau commented:
@afedchin Good to know that this situation doesn't appear on each users system who has multiple versions installed. I hope that the reinstall today solves the issue, If not, I'll go and buy a system only for kodi compile. Reinstalling my windows is currently not possible....

Razzeee commented:
Compiled fine for me having 2013 Premium and 2015 Enterprise installed.
Had to download the dependencies a few times but other than that pretty painless.

mapfau commented:
I'm happy with you really :-)

I finaly copied the ucrt header and lib files from sdk10 into the sdk 8.1 kit and now it starts to get better.

- I dropped all things in xbmc/project, did a git reset --hard to pill out the deleted things from the git repo (its a rebased xbmc master in my local repo) This ends up in files and folders with  really strange security rights, I was not able to write / execute anythin in it. chmod 775 on the complete folder fixes it.

- Now after calling DownloadBuildDeps.bat I have the following (german) message:

```
--------------------------------------------------
Extracting zlib-1.2.8-win32-vc140-v2.7z...

Die Datei kann nicht in sich selbst kopiert werden.
        0 Datei(en) kopiert.

==================================================
Some formed packages had errors


bzip2-1.0.5-win32.7z: - Can't extract files from archive bzip2-1.0.5-win32.7z

ERROR: Not all formed packages are ready!
```

I bet nobody else have this problems, right? I just post it here to be sure.....

garbear commented:
try cloning the repo fresh. I always have problems when throwing git fu at the repo, and I've found the only 100% way is to start fresh.

To save time, you can clone from a local source. Rename your repo folder (e.g. "kodi-bak") and then clone it
```
git clone kodi-bak kodi
```
you'll have to redirect the origin remote
```
cd kodi
git remote rm origin
git remote add origin git@github.com:mapfau/xbmc.git
```

phil65 commented:
```
git clean -xfd
```
always worked for me up to now, also in this case.

garbear commented:
git clean -xdf doesn't remove cloned repos. I've had these stale repos cause problems

mapfau commented:
I stop now with this adventure and wait some weeks until anything is proved. Sorry, its not working at all for me anymore. I pulled again the repo directly from kodi (removed anything before). I cannot run DownLoadBuildDeps.bat. Permission denied. There is a user SID_NULL in my securitiy settings for this and all files in project, Windows claims that users are unsorted and so on.
I spend now about 60 hours on this. I stop and follow this thread until I'm sure it will work again.

Paxxi commented:
If you get permission issues it something with your machine most likely and this PR's accidentally surfaced those issues, it's not something that we will be able to fix in the repo as nothing in our build system touches any permissions. 
People have had issues when they accidentally run a command such as git clone downloadbuilddeps causing weird issues but it's usually resolved by wiping the repo. 

As you mentioned that reinstall is not an option maybe spinning up a vm can be a workaround until you can troubleshoot more or reinstall. 

If you want help troubleshooting windows I'll gladly help, can you post the full error messages and error codes you get? Or email me directly at username at Kodi dot tv

mapfau commented:
thanks @paxxi, I'll try to figure it out at work with some of my collegiues. Maybe I have to come back to you. It really seems that my system is angry about all the SDK and VS reinstalls and uninstalls.

mapfau commented:
Just a question: I tried 4 times now make-mingwlibs.bat.
ffmpeg works fine, the folloing three always fail:

```
-------------------------------------------------
 building libdvd 32bit
-------------------------------------------------
backgrounding: sh ./buildlibdvd.sh clean /xbmc/project/Win32BuildSetup/bgprocess                                                                 mingw & (workdir: /xbmc/tools/buildsteps/win32)
waiting on bgprocess...
../buildhelpers.sh: Zeile 138: git: Kommando nicht gefunden.
libdvdcss-master.tar.gz ...................................... [Downloading]
--2016-04-05 10:53:58--  https://github.com/xbmc/libdvdcss/master.tar.gz
Auflösen des Hostnamens »github.com (github.com)« … 192.30.252.130
Verbindungsaufbau zu github.com (github.com)|192.30.252.130|:443 … verbunden.
HTTP-Anforderung gesendet, auf Antwort wird gewartet … .......404 Not Found
2016-04-05 10:54:34 FEHLER 404: Not Found.

```
Only on my machine again?

Montellese commented:
Looks like @mapfau is not the only one with problems, see http://forum.kodi.tv/showthread.php?tid=267372. Looks like the nightly builds don't work/run anymore.

mapfau commented:
Just found in buildffmpeg.sh:

  ```
do_addOption "--enable-debug"
  do_addOption "--toolchain=msvc"

  PATH="/c/Program Files (x86)/Microsoft **Visual Studio 12.0**/VC/BIN/":$PATH
```

shouldn't this be 14.0 ?

mapfau commented:
I reinstalled git now, and it seems to work, sorry for the interuption...

mapfau commented:
Is DVD working on any of your system?

 ```
	Kodi.exe!CEmuFileWrapper::GetFileXbmcByStream(_iobuf * stream) Zeile 220	C++
 	Kodi.exe!dll_vfprintf(_iobuf * stream, const char * format, char * va) Zeile 1571	C++
 	Kodi.exe!dll_fprintf(_iobuf * stream, const char * format, ...) Zeile 1625	C++
 	libdvdnav.dll!63a94f8a()	Unbekannt
 	[Unten angegebene Rahmen sind möglicherweise nicht korrekt und/oder fehlen, keine Symbole geladen für libdvdnav.dll]	
 	libdvdnav.dll!63a951c2()	Unbekannt
 	[Externer Code]	
>	Kodi.exe!DllDvdNav::dvdnav_open(dvdnav_s * * p1, const char * p2) Zeile 256	C++
 	Kodi.exe!CDVDInputStreamNavigator::Open() Zeile 100	C++
 	Kodi.exe!CVideoPlayer::OpenInputStream() Zeile 793	C++
	Kodi.exe!CEmuFileWrapper::GetFileXbmcByStream(_iobuf * stream) Zeile 220	C++
 	Kodi.exe!dll_vfprintf(_iobuf * stream, const char * format, char * va) Zeile 1571	C++
 	Kodi.exe!dll_fprintf(_iobuf * stream, const char * format, ...) Zeile 1625	C++
 	libdvdnav.dll!63a94f8a()	Unbekannt
 	[Unten angegebene Rahmen sind möglicherweise nicht korrekt und/oder fehlen, keine Symbole geladen für libdvdnav.dll]	
 	libdvdnav.dll!63a951c2()	Unbekannt
 	[Externer Code]	
>	Kodi.exe!DllDvdNav::dvdnav_open(dvdnav_s * * p1, const char * p2) Zeile 256	C++
 	Kodi.exe!CDVDInputStreamNavigator::Open() Zeile 100	C++
 	Kodi.exe!CVideoPlayer::OpenInputStream() Zeile 793	C++

```

the iobuf stream is a NULL pointer on my system (this dvd works before well)

afedchin commented:
> shouldn't this be 14.0 ?

it's not related and used for debugging purpose only

mapfau commented:
One other thing makes me currently some headache:

loading the plugin.video.prime addon (I already deinstalled / installed the plugin) shows me in the log:

```
dependent on version 2.1.0 of the xbmc.python api
19:47:36 T:9040   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.ImportError'>
                                            Error Contents: DLL load failed: Die angegebene Prozedur wurde nicht gefunden.
                                            Traceback (most recent call last):
                                              File "C:\Users\Markus\AppData\Roaming\Kodi\addons\plugin.video.prime_instant-fixes\default.py", line 3, in <module>
                                                import urllib
                                              File "C:\Dev\xbmc\system\python\Lib\urllib.py", line 26, in <module>
                                                import socket
                                              File "C:\Dev\xbmc\system\python\Lib\socket.py", line 47, in <module>
                                                import _socket
                                            ImportError: DLL load failed: Die angegebene Prozedur wurde nicht gefunden.
                                            -->End of Python script error report<--
```

Help really appreciated- if solved I can continue with skygo :-)

Voyager1 commented:
same here. DVD playback doesn't work (at least iso files, haven't tried anything else yet). Seems like all stream pointers have a null _Placeholder member coming from libdvdnav.dll

Paxxi commented:
The file object has been made opaque with vs2015, it only exposes one member and that's _Placeholder. Any file object not opened through the emulation layer will have it set to null. 

As there's several issues caused by this pr I'm trying to work through them in order of priority which currently is
1. Fix python (found solution, needs testing and proper packaging, hopefully done Friday) 
2. Curl ssl
3. Libdvd

Unfortunately I have a planned after work today so best estimate is solutions start appearing tomorrow unless someone else can chip in
Again, sorry for all the trouble.

Montellese commented:
Has anyone tried the unit tests yet? I get some errors that it can't find some DLLs because they are missing from the `DelayLoadDLLs` VS property After manually copying them it starts but immediately crashes but I haven't figured out why yet.

mapfau commented:
Unfortunately there is also an issue vith the opaque FILE in emu_msvcrt.h

  FILE* f = fopen("C:\\Temp\\wrm.xml", "wb");
  fwrite(xml_start, 1, xml_size, f);
  fclose(f);

fwrite doesn't write into file, because it is detected as STDIN or STDOUT in the macros:

#define IS_STDOUT_STREAM(stream)    (stream != NULL && __IS_STDOUT_STREAM(stream))
#define IS_STDERR_STREAM(stream)    (stream != NULL && __IS_STDERR_STREAM(stream))

Voyager1 commented:
@Paxxi and all - I have a working solution for libdvd! I just need to make it cross-platform and I'll send a patch later today.

Voyager1 commented:
fix for libdvd and vs2015 here : https://github.com/Voyager1/xbmc/commit/c53ccdd2e127bf50a75495e0e0b593c150bb1632

Voyager1 commented:
@mapfau I also had to assume in the libdvd fix (above) that a file defined and created in libdvd has to be stdout or stderr. But that's ok in my case.

mapfau commented:
@Voyager1 means that part of your fix could be used in emu_msvccrt also?
I have recognised yesterday, that std::ofstream is also not working when using in addons.
It leads to a segfault - but similar thing stream = NULL when trying to write things.

I'll look over your patch asap... thx

Voyager1 commented:
@mapfau - you can't fill the placeholder if you didn't create the file object, like in the case of the stdout/err definition inside libdvd. When otoh libdvd comes back to the emu layer to open a file inside our vfs, that's where we control file object creation.

Voyager1 commented:
I have added another patch which doesn't fix anything more, but safeguards all FILE* access in the emu wrapper. Also added the check for MSVC version which was missing.
https://github.com/Voyager1/xbmc/commit/cba65c7b7741ba7d814b52bb9d55dcdcb3765255

fritsch commented:
The IsValidStream is only working for windows as the placeHolder is nullptr
for all the other platforms, right? And making it general for other
platforms would mean a lot of #ifdeffery.

Can't we solve this properly without the ifdeffery by using basic OO
methods?

2016-04-08 10:25 GMT+02:00 Voyager1 <notifications@github.com>:

> I have added another patch which doesn't fix anything more, but safeguards
> all FILE* access in the emu wrapper. Also added the check for MSVC version
> which was missing.
> Voyager1@cba65c7
> <https://github.com/Voyager1/xbmc/commit/cba65c7b7741ba7d814b52bb9d55dcdcb3765255>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/8992#issuecomment-207308799>
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

Voyager1 commented:
@fritsch - the _Placeholder member does only exist on Win32/VS2015. It's MS way to make the FILE object opaque. I don't see how you can avoid ifdef's here. Just take a look at "IsValidStream", it's an expression that checks the pointer not being null and only on Windows/VS2015 it checks also the platform-specific _Placeholder.

fritsch commented:
Yeah seen after looking at the whole class. 

That file now has two parts and also nearly two implementations separated by #ifdefs between windows and the rest. Also when you look at the memset to reset all the fields, this implementation is now also special for windows, where we don't guarantee anymore that all fields are zeroed out, when removing a Fiel from the internal list. Also the list implementation which always causes O(n) to reuse a File / get a new one is not state of the art.

My points is: If we touch and discuss that file now, we should fix those design issues right away.

Voyager1 commented:
@fritsch the thing about O(n) always when reuse a file is incorrect. It's only upon start and end when we create a pool of emulated files. And there's really no other way as the way a File* now works since they made it opaque in VS2015 is that there is only that _Placeholder which hides the internals. Now we don't care what it is (since we emulate anyway) but it has to have a value to distinguish it from the ones not created by our emulation layer (hence the need to assign an actual address pointing to something). I don't think the way @paxxi did this can be done better.

fritsch commented:
Perhaps a good time to rethink our emulation approach and kill it? :-)

Voyager1 commented:
@fritsch - we need emulation to fool libdvd (and other libraries) to make them think they're handling real files. No emulation means we can't use those libs anymore without major rewrites. My opinion.

Voyager1 commented:
Further cleaned up code, made it more "complete" and raised PR #9583

mapfau commented:
@Voyager1 To be honest, I'm completely lost with all the changes regarding VS2015 / CMakeLists done in the last few days. Could be pls. anybody so kind and can answer my two main questions?

- why can the emulation approach not be bound only to dvdplayer. Every Addon and I write currently runs through this emulation methods (also all dlls I load from my binary addons. @Voyager1 Do I understnd right that you have a working approach wichout emulation layer?

- has there any fix regarding the STDIN / STDERR in the emulation layer be pushed into the kodi master tree? I have a lot of trouble with encryption of media streams because there are modules wich has to write files to work properly.

Thanks, sorry for not beeing able to follow this strong speed here :-)

Voyager1 commented:
@mapfau - there is currently no working approach for libdvd without emulation layer. While the version we use potentially (theoretically) allows the use of streaming dvds, I haven't figured out how yet. It would maybe allow us to get rid of emulation, probably with some limitations (inability to read physical dvds anymore i'm afraid).

What I did for stdin/out/err in #9583 is make the assumption that an external lib such as libdvd defines std streams and as such the FILE* coming into the emulation layer are not recognized and 'invalid' (due to opaque type). However the fix I did, ensures that output to these streams is properly written into our logs. There's (currently) no better fix for that. If you work with streams opened through our emu layer, the addons should work fine.

mapfau commented:
thanks, @Voyager1. Then I'll pull the latest kodi now.

libdvd is compiled during the kodi dependency stuff, right? Why isn't it possible to link the msvcrt_emu stuff only für this libdvd library. Is there a reason why the emulation is placed at such near to kodi and therefore the other addons (just want to understand the context about it)

Voyager1 commented:
Emulation is the way into our kodi VFS for all the external libraries that only know the standard MSVCRT API. I think there are other libraries that need it although I'm not sure. The file wrapper has been around since long before I joined. There are folks like @akva2 who know a lot more about this.

FernetMenta commented:
I suggest to apply workarounds only to those parts that need it. That would make the design much cleaner and more resilient.

akva2 commented:
on windows there is no need to link anything to use the emulation. they are runtime overloaded on dll load. on osx/linux it's done through a linker map, and thus we need to link specific versions the libraries (with the linker map and the wrapper functions). the wrapper functions call the kodi emulation functions, which are undefined in the binary (but around when kodi loads the shared library).

mapfau commented:
@akva2 thanks, Not totally clear for me wich part of kodi / cpluff does the overloading, but if there is some time I'll try to figure it out....

akva2 commented:
overloading is in xbmc/cores/DllLoader/&lt;windows related files&gt;

FernetMenta commented:
@mapfau https://github.com/xbmc/xbmc/blob/master/xbmc/cores/DllLoader/Win32DllLoader.cpp#L39

As a first step I may make sense to align win32 with Linux and link against the wrapper.

akva2 commented:
that makes little sense. you need a specific wrapper for windows cause symbols are mangled differently and more symbols are overridden. furthermore the windows linker cannot do link-time maps in the same way.

FernetMenta commented:
Then use a specific wrapper for the specific cases and not for all cases. In a first iteration we would have two loaders for Windows. The specific loader can be moved to an addon later.

akva2 commented:
what specific wrapper case ?

FernetMenta commented:
sorry, I meant dll loader.

Voyager1 commented:
fyi - since 5.0.3 libdvd supports opening streams. I'll take a look into it, it does not appear to be difficult.
http://git.videolan.org/?p=libdvdnav.git;a=commitdiff;h=48102c2a5252e2e4ce3e2d71154c79ba8e339de9

mapfau commented:
wow, this would make emulation obsolete! And the discussion here :-)

Voyager1 commented:
don't get too optimistic. For reading ISO files this is relatively simple (only 3 callback functions are needed, read, readv and seek). ISOs can be given to libdvd as a single stream. I don't see how you can turn a DVD device on Windows or (even worse) a VIDEO_TS folder & files structure into a single stream object. It's kind of the reverse of parsing a UDF structure, we would have to simulate a UDF structure around VIDEO_TS folder/files... ideas?

akva2 commented:
On 12/04/16 15:09, Voyager1 wrote:
>
> don't get too optimistic. For reading ISO files this is relatively 
> simple (only 3 callback functions are needed, read, readv and seek). 
> ISOs can be given to libdvd as a single stream. I don't see how you 
> can turn a DVD device on Windows or (even worse) a VIDEO_TS folder & 
> files structure into a single stream object. It's kind of the reverse 
> of parsing a UDF structure, we would have to simulate a UDF structure 
> around VIDEO_TS folder/files... ideas?
>
yell at the dvdread authors ;) the first should be kinda trivial - it's 
a do-nothing condition (you just use the dvdread handling of it and 
bypass the vfs). the second is a bitch. i tried to use this a year ago 
or something and then realized this limitation. you have to generate an 
udf-fs on the fly. it *is* doable - cdrecord (or at least one of the 
numerous forks) can do it, but i got fed up and moved on. that would be 
where to dig for code.

ace20022 commented:
Couldn't it be easy to patch dvdread/libdvdnav?

akva2 commented:
that's the alternative. that was under the limitation of vanilla dvd libs.

garbear commented:
You could take a page from the book of the RetroArch/libretro team. Over 20 (!) of their cores have VFS support because they patched in a "read from vfs" capability.

akva2 commented:
Or from 60% of my audio decoders..

-------- Opprinnelig melding --------
Fra: Garrett Brown <notifications@github.com>
Dato: 12.04.2016 22:17 (GMT+01:00)
Til: xbmc/xbmc <xbmc@noreply.github.com>
Ko: Arne Morten Kvarving <Arne.Morten.Kvarving@sintef.no>
Emne: Re: [xbmc/xbmc] [win32][xbmc] Upgrade to VS2015 (#8992)


You could take a page from the book of the RetroArch/libretro team. Over 20 (!) of their cores have VFS support because they patched in a "read from vfs" capability.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://github.com/xbmc/xbmc/pull/8992#issuecomment-209084881>

Voyager1 commented:
> it *is* doable - cdrecord (or at least one of the
> numerous forks) can do it, but i got fed up and moved on. that would be
> where to dig for code.

Honestly, I don't think I have the bandwidth to do this. We will have to call for a decision between 2 options A) continue with emulation as is, or B) drop emulation but on no longer supporting DVD from VIDEO_TS/ VOB/IFO files. I'm (selfishly) OK with this, as I only use ISO images or physical device anyway... Perhaps this is something we take to the forum and drive to a decision there.
